uber/models.py

```diff
@@ -1625,6 +1817,464 @@ def must_contact(self):
+            self.uses_allowed = None
+
+        # If 'discount' is empty, then this is a full discount, free badge
+        if not self.discount:
+            self.discount = None
+
+        self.code = '' if self.code is None else self.code.strip()
+        if not self.code:
+            # If 'code' is empty, then generate a random code
+            self.code = self.generate_random_code()
+        else:
+            # Replace multiple whitespace characters with a single space
+            self.code = re.sub(r'\s+', ' ', self.code)
+
+    def calculate_discounted_price(self, price):
+        """
+        Returns the discounted price based on the promo code's `discount_type`.
+
+        Args:
+            price (int): The badge price in whole dollars.
+
+        Returns:
+            int: The discounted price. The returned number will never be
+                less than zero or greater than `price`. If `price` is None
+                or a negative number, then the return value will always be 0.
+        """
+        if not self.discount or not price or price < 0:
+            return 0
+
+        discounted_price = price
+        if self.discount_type == self.FIXED_DISCOUNT:
+            discounted_price = price - self.discount
+        elif self.discount_type == self.FIXED_PRICE:
+            discounted_price = self.discount
+        elif self.discount_type == self.PERCENT_DISCOUNT:
+            discounted_price = int(price * ((100.0 - self.discount) / 100.0))
+
+        return min(max(discounted_price, 0), price)
+
+    @classmethod
+    def _generate_code(cls, generator, count=None):
+        """
+        Helper method to limit collisions for the other generate() methods.
+
+        Arguments:
+            generator (callable): Function that returns a newly generated code.
+            count (int): The number of codes to generate. If `count` is `None`,
+                then a single code will be generated. Defaults to `None`.
+
+        Returns:
+            If an `int` value was passed for `count`, then a `list` of newly
+            generated codes is returned. If `count` is `None`, then a single
+            `str` is returned.
+        """
+        with Session() as session:
+            # Kind of inefficient, but doing one big query for all the existing
+            # codes will be faster than a separate query for each new code.
+            old_codes = set(s for (s,) in session.query(cls.code).all())
+
+        # Set an upper limit on the number of collisions we'll allow,
+        # otherwise this loop could potentially run forever.
+        max_collisions = 100
+        collisions = 0
+        codes = set()
+        while len(codes) < (1 if count is None else count):
+            code = generator().strip()
+            if not code:
+                break
+            if code in codes or code in old_codes:
+                collisions += 1
+                if collisions >= max_collisions:
+                    break
```

Comment:
If we hit max_collisions, the only thing that happens is that we fail to create as many promo codes as requested.

Consider the case where someone requests a thousand promo codes that are only one character long. We can't possible create that many unique one-character promo codes. We could calculate the total number of possible promo codes that we can generate, and if the requested number is larger than that, just generate all of them.

But what if the user requests 99 promo codes, and the keyspace allows for 100? Do we still randomly generate 99, or do we generate 100 and randomly remove one? What if the user requests 50 promo codes and the keyspace allows for 100? There's a lot of corner cases to do this efficiently.

So I chose the stupid way of doing it! We do our best to meet the users request, but hey, if we can't do it we gotta give up at some point! 100 is just the arbitrary point I chose (there's probably some statistics formula to figure out the optimal number of collisions, but again that seems overwrought).